src/main/java/com/avispl/symphony/dal/communicator/crestron/CrestronXiO.java

```diff
@@ -16,119 +23,204 @@
+     */
+    private Map<String, AggregatedDevice> aggregatedDevices = new ConcurrentHashMap<>();
 
     /**
      * Executor service for handling background tasks
      */
     private ExecutorService executorService;
 
-	/**
-	 * Interceptor for RestTemplate that injects
-	 * authorization header and fixes malformed headers sent by XIO backend
-	 */
-	private ClientHttpRequestInterceptor xioHeaderInterceptor = new CrestronXioHeaderInterceptor();
+    /**
+     * Interceptor for RestTemplate that injects
+     * authorization header and fixes malformed headers sent by XIO backend
+     */
+    private ClientHttpRequestInterceptor xioHeaderInterceptor = new CrestronXioHeaderInterceptor();
+
+    /**
+     * Instance of device data loader worker
+     */
+    private CrestronXioDeviceDataLoader deviceDataLoader;
+
+    /**
+     * This parameter holds timestamp of when we can perform next API request for retrieving devices metadata (device list)
+     */
+    private long nextDevicesListCallTs;
+
+    /**
+     * This parameter holds timestamp of when we can perform next API request for retrieving device statistics
+     * So the maximal available fetch rate is utilized (5 devices per second)
+     */
+    private volatile long nextDeviceStatusCallTs;
+
+    /**
+     * This parameter holds timestamp of when we need to stop performing API calls
+     * It used when device stop retrieving statistic. Updated each time of called #retrieveMultipleStatistics
+     */
+    private long validRetrieveStatisticsTimestamp;
+
+    /**
+     * Time period within which the device metadata (basic devices information) cannot be refreshed.
+     * If ignored if device list is not yet retrieved or the cached device list is empty {@link CrestronXiO#aggregatedDevices}
+     */
+    private long validDeviceMetaDataRetrievalPeriodTimestamp;
+
+    /**
+     * Aggregator inactivity timeout. If the {@link CrestronXiO#retrieveMultipleStatistics()}  method is not
+     * called during this period of time - device is considered to be paused, thus the Cloud API
+     * is not supposed to be called
+     */
+    private long retrieveStatisticsTimeOut = 3 * 60 * 1000;
 
-	/**
-	 * Instance of device data loader worker
-	 */
-	private CrestronXioDeviceDataLoader deviceDataLoader;
+    /**
+     * Device metadata retrieval timeout. The general devices list is retrieved once during this time period.
+     */
+    private long deviceMetaDataInformationRetrievalTimeout = 30 * 60 * 1000;
 
-	/**
-	 * This parameter holds timestamp of when we can perform next API request
-	 */
-    private long nextApiCallTs;
+    /**
+     * Indicates whether a device is considered as paused.
+     * True by default so if the system is rebooted and the actual value is lost -> the device won't start stats
+     * collection unless the {@link CrestronXiO#retrieveMultipleStatistics()} method is called which will change it
+     * to a correct value
+     */
+    private boolean devicePaused = true;
```

Comment:
This might be a candidate to be volatile - unless it is always used under the lock